### DISCUSSION TITLE
On-Demand Paging usage specified by the application

### DISCUSSION THREAD

**Discussion by swelch** [ofiwg/libfabric#9144] [Q&A]
While the FI_MR_ALLOCATED mode bit indicates that the provider requires that registered memory regions must be backed by physical memory pages at the time the registration call is made, it is not clear to me in libfabric man pages how ODP optional usage should be indicated for a provider that supports it. For instance, if a provider performs RDMA and can be configured to support ODP, how should this be selected by the application in a libfabric common way. My assumption is that the application and provider can let the presence of the FI_MR_ALLOCATED mr_mode bit select whether ODP is desired or not. To support this case the provider may export multiple fi_info per domain, one with FI_MR_ALLOCATED MR mode bit set and one without it set. The application would select the correct fi_info based on their mr_mode hints they provide. 

It appears the existing Verbs provider code requires that an undocumented environment variable "FI_VERBS_USE_ODP" be set to enable ODP, that the verbs extended query capabilities API is supported, that the verbs hardware supports ODP, and that the registration is for system memory. If this is true it will pass the appropriate attribute to the rdma-core. Independent of these settings it still requires FI_MR_ALLOCATED mr_mode bit.

@shefty - Thoughts on the correct usage and/or the best way an application can select that it wants to use ODP with an RDMA provider that supports ODP?

**Comment by shefty** [comment]
I would start by defining how ODP is being used.  I believe the verbs provider enables it through an environment variable because having it enabled gives lower performance than some alternatives.

For buffers that are only accessible locally (e.g. send and receive buffers), ODP could eliminate any MR cache.  The FI_MR_LOCAL bit could be cleared for this use case, with memory registration hidden from the app completely.

If we consider remotely accessed buffers (e.g. remote RMA read/write), I agree ODP aligns with FI_MR_ALLOCATED.  This also ties into the MR cache and invalidating entries.

I would like to leave it up to the provider to determine whether enabling ODP provides better performance, versus using a MR cache.  I don't think we get away from having some sort of environment variable to specify whether using ODP and/or the MR cache is acceptable from a security standpoint.

**Reply by swelch** [reply]
@shefty - This seems reasonable. Currently we export and fi_info with FI_MR_ALLOCATED in the mr_mode before an fi_info without FI_MR_ALLOCATED since non-ODP would be preferred. The thought was that fi_info query without hints would display both fi_info making it clear both are supported. However, using an environment variable to turn on ODP and control not setting FI_MR_ALLOCATED in the mr_mode seems like a suitable option as well. I'll discuss internally and migrate to that usage if possible.

**Reply by shefty** [reply]
We should have a broader discussion, maybe as part of 2.0, about whether the existence of a MR cache should be exposed to the application through the API.

**Comment by chuckfossen** [comment]
[ODP presentation for OFIWG.pdf](https://github.com/ofiwg/libfabric/files/15476379/ODP.presentation.for.OFIWG.pdf)
I updated the presentation based on feedback from today's meeting. The section on FI_MR_LOCAL updates may need some more refinement.

**Reply by shefty** [reply]
The presentation doesn't indicate what the actual problem is that it's trying to solve.  The closest is: "forcing ODP for all memory registrations is wrong", without going into detail.

ODP is a HW/provider implementation detail for dealing with VA to PA mapping changes.  There are other possible solutions which do not involve ODP.  From the viewpoint of the app, it needs to know if the region passed into registration must be backed by physical pages -- i.e. FI_MR_ALLOCATED.  Implicitly, FI_MR_ALLOCATED has also indicated support for dynamic VA to PA mapping changes.  That could be called out explicitly, possibly even adding an alias name for FI_MR_ALLOCATED (e.g. FI_MR_DYNAMIC).

To disable dynamic mapping during registration, passing in a simple flag would work. (With FI_MR_ALLOCATED = 0, the call assumes dynamic mapping by default.)

FI_MR_ALLOCATED set to 0 requires that RDMA transfers that target the region work as expected.  The RDMA should update the physical pages backing the region.  If no physical page exists, the transfer completes in error.  That's assumed by the current definition.
 
Memory registration is covered by a specific set of MR bits.  Don't include MR related bits with non-MR caps and mode bits.  MR bits are specifically isolated because everything about memory registration is horrible.

I would not expose ODP as a generic domain attribute.  At best make it a setopt, so developers can ignore it like all other options.  But even then, it's still exposing implementation details through the API.  Control over the MR cache hasn't even been exposed yet.

Changing how the desc parameter is used when null vs non-null seems reasonable.  Though I don't see why an app with FI_MR_LOCAL disabled would ever bother with the overhead of registering a local buffer to pass in a valid descriptor.  From the viewpoint of the app, it's taking an action which should hurt its performance (registering memory when it was told it didn't need to).  If the provider really wanted local buffers registered, and the app indicated that it can do that, then the provider shouldn't have cleared the bit.

**Reply by iziemba** [reply]
Slide 2 lists the 4 behaviors HPE needs to support with libfabric. These behaviors are from the use cases our stakeholders are asking for.

I can see where you are coming from in stating that ODP is a HW/provider implementation detail. If we can come up with an optimal way to support page migration, fork, dynamic MRs, and THP without exposing ODP to users, that would be great.

The issue we currently have is the provider has no idea if page migration or dynamic/sparse VA range will be used (I am going to ignore fork and THP for now). These are all application properties. This will require providers to adapt to changes in the VA to PA mapping. Since not all providers may be able to support this, the application now needs to know if the provider can handle dynamic VA to PA mappings. We need a mechanism for application and provider to negotiate/discover this. I am ok with this being called dynamic instead of ODP.

Updating FI_MR_ALLOCATED to support dynamic behavior, at face value, does seem not right. Today, verbs and CXI provider implement FI_MR_ALLOCATED by page pinning. Forcing this dynamic mapping to be the default behavior would require verbs and CXI provider to implement ODP.

At face value, FI_MR_ALLOCATED being zero, meaning always dynamic, does seem to work for supporting page migration and dynamic/sparse VA ranges.

This seems to leave us in a place where either a new dynamic flag needs to be defined for FI_MR_ALLOCATED (or something equivalent), or apps need to update to support FI_MR_ALLOCATED being zero.



> Changing how the desc parameter is used when null vs non-null seems reasonable. Though I don't see why an app with FI_MR_LOCAL disabled would ever bother with the overhead of registering a local buffer to pass in a valid descriptor.

Some applications register remote MRs for large chunks of persistent memory (e.g. SHMEM heap) but do not want to implement the tracking of MRs outside this region. Changing the description of desc param without FI_MR_LOCAL sounds good. Thinking out loud.... if we make this the default behavior (yes it would put onus of verbs and other providers to query desc param and do internal memory registration), we could potentially eliminate FI_MR_LOCAL and FI_MR_HMEM.

**Reply by shefty** [reply]
How would an application change its behavior WRT a new FI_MR_ODP mode bit, versus FI_MR_ALLOCATED?

For FI_MR_ALLOCATED, I'm saying that it implies that pages may change dynamically. because no pages even need to exist when register is called.  The name is just confusing.  FI_MR_ALLOCATED = 1 tells the app not to do anything *explicitly* (e.g. free() then malloc()) that changes the VA to PA mapping.  FI_MR_ALLOCATED = 0 allows the app to change the mapping.  AFAICT, these are the 2 semantics that seem useful.  What semantic is missing?

If an app supports FI_MR_ALLOCATED = 1, then they naturally support FI_MR_ALLOCATED = 0.

Note that the app is unaware of page pinning.  That, too, is a provider detail.  Page migration that occurs outside of the scope of the app becomes a provider issue.  E.g. kernel HMEM migrating buffers from host to device. There's little an app can do about that.  I would expect the provider to either integrate with page migration or disable migration (via pinning).

If an app pulls buffers from a pool for sends and receives, it's trivial for the app to register the buffers.  It's more difficult and costly for the provider to do the same, since it's unaware that a pool is being used, or that a single registration might work for the entire pool, versus 1 registration per buffer.  That's why we want to keep FI_MR_LOCAL.

**Reply by iziemba** [reply]
>For FI_MR_ALLOCATED, I'm saying that it implies that pages may change dynamically. because no pages even need to exist when register is called.

The following is from https://ofiwg.github.io/libfabric/v1.21.0/man/fi_mr.3.html.

**FI_MR_ALLOCATED**
_When set, all registered memory regions must be backed by physical memory pages at the time the registration call is made._

This seems to go against your claim of no pages need to exist when register is call, right?

>Note that the app is unaware of page pinning. That, too, is a provider detail. Page migration that occurs outside of the scope of the app becomes a provider issue. E.g. kernel HMEM migrating buffers from host to device. There's little an app can do about that. I would expect the provider to either integrate with page migration or disable migration (via pinning).

The use-case a lot of GPU vendors are looking to support is being able to use a malloc() buffered with GPU kernels. Depending on hardware platform, this may result in zero or many page migrations when RDMA is done against corresponding VA. Apps need to know if a provider can support this. Not all providers may be able to support this. How are apps suppose to know this?

> If an app pulls buffers from a pool for sends and receives, it's trivial for the app to register the buffers. It's more difficult and costly for the provider to do the same, since it's unaware that a pool is being used, or that a single registration might work for the entire pool, versus 1 registration per buffer. That's why we want to keep FI_MR_LOCAL.

Allowing users to pass in a valid desc parameter solves this problem. HPE has done a CXI provider specific extension to enable this, and it has been working well.

**Reply by shefty** [reply]
If FI_MR_ALLOCATED = *0*, then no pages need to exist.  It's only when the flag is set that the pages must be present.  I'm saying what's being implied when FI_MR_ALLOCATED = 1 is that the app cannot replace the mapping.  We should explicitly define that additional restriction.

Automatic page migration between host to GPU buffers is what I'm referring to with the kernel hmem driver.  That level of migration is outside the control of the app.  What I'm looking for is, if a provider doesn't hook into the kernel hmem support, what does the app *do* with that information?  When an FI_MR_xxx bit is set, it means that the app must do something different in its code.

In this case, the app called malloc() or another OS call, got back a buffer, and later the GPU accessed the buffer.  The OS, deciding to be helpful, migrated some pages backing the buffer.  What are we asking the app to do?  Not access any buffers it allocates from the GPU?  Not register any memory regions allocated using an OS call which are later accessed by the GPU?  There's a more abstract issue here unrelated to ODP.

So far, I think we agree on adding some flag to register memory for the app to indicate that *the app* won't change the VA mapping.  We also agree that the app could pass in a valid desc for a local buffer if it's registered the memory.  But what we don't have in the latter case is a way for the app to know that doing this might be useful.  If FI_MR_LOCAL = 0, most apps aren't going to register the buffers in the first place.

**Reply by iziemba** [reply]
>If FI_MR_ALLOCATED = 0, then no pages need to exist. It's only when the flag is set that the pages must be present. I'm saying what's being implied when FI_MR_ALLOCATED = 1 is that the app cannot replace the mapping. We should explicitly define that additional restriction.

I agree that FI_MR_ALLOCATED = 1 should have this additional restriction explicitly defined.

>In this case, the app called malloc() or another OS call, got back a buffer, and later the GPU accessed the buffer. The OS, deciding to be helpful, migrated some pages backing the buffer. What are we asking the app to do? Not access any buffers it allocates from the GPU? Not register any memory regions allocated using an OS call which are later accessed by the GPU? There's a more abstract issue here unrelated to ODP.

If the app sees that FI_MR_ALLOCATED = 0, the app/middleware does not have to do anything different. But if FI_MR_ALLOCATED = 1, the app/middleware may not be able to use malloc() memory for RDMA and be accessible from the GPU. This type of issue has existed for a while. For example, cudaManagedMemory typically cannot support RDMA. Thus, apps/middleware may have to use bounce buffers.

>So far, I think we agree on adding some flag to register memory for the app to indicate that the app won't change the VA mapping.

Is a new flag needed if the FI_MR_ALLOCATED = 1 documentation is updated to state that VA <-> PA mapping must not change while buffer is being used for RDMA or explicitly registered?

>We also agree that the app could pass in a valid desc for a local buffer if it's registered the memory.

I agree.

>But what we don't have in the latter case is a way for the app to know that doing this might be useful. If FI_MR_LOCAL = 0, most apps aren't going to register the buffers in the first place.

The best I think we can do here is provide guidance. Apps are encouraged to explicitly register memory and provide a valid MR desc for local data transfers. By not doing this, the provider may have to perform internal memory registration. This may or may not hurt performance. For example, the provider may need to track the the memory registration. This could result in the provider having to internally enable completion events, even if FI_SELECTIVE_COMPLETION is used, to ensure forward progress.

**Reply by shefty** [reply]
The app doesn't know that the pages that it allocated through malloc() might end up on the GPU!  **What are we asking the app to do?**

It called malloc(), registered the returned region, and requested an RDMA to that region.  Are we telling the app not to use malloc()?  Never perform RDMA?  Never access a buffer from the GPU which was allocated using malloc(), because that *might* migrate the page?

I still don't know what an app is supposed to do with this information.  The app didn't perform the page migration to/from the host to the device.  The OS is doing that on a whim.  How does the app know that a bounce buffer isn't going to migrate to the GPU?  I mean, after it performs an RDMA, it's going to copy data from the bounce buffer to a GPU buffer.  From the viewpoint of the OS, the only access to the bounce buffer is by the GPU, so why wouldn't it migrate the bounce buffer too?  How does the app prevent it?

**Reply by iziemba** [reply]
If OS triggers memory migration, there is nothing the provider can do without ODP. This is why, in HPC, things which cause the OS to migrate memory (THP, NUMA balancing, etc...) are explicitly disabled. If ODP is disabled and memory migration happens, depending on provider/driver MR implementation, either RDMA operations will unexpectedly fail or data corruption will occur (stale pages will be referenced).

We now have hardware platforms where host <-> GPU memory migration is supported for malloc() buffers and users are being encouraged to use this functionality. Specifically, HPE stakeholders are asking for RDMA and migratable GPU memory support. From a CXI provider point of view, this requires ODP to avoid unexpected RDMA failure.

But, if I am an app/middleware and running on hardware platform where host <-> GPU memory migration may happen, I need to know if the provider can handle to this avoid RDMA operation failure or data corruption.

In addition, updating FI_MR_ALLOCATED = 1 documentation to state that page migration is supported seems wrong because this now forces providers to support ODP or the equivalent.

**Reply by shefty** [reply]
We need to define something **actionable** on the part of the app.  Example:

```
FI_MR_BIT_5
When set, the memory region must be backed by physical pages when registered.
While the region is registered, the app must not free or remap the region.
```
The action here is that the app must call malloc(), calloc(), mmap(), cudaMalloc(), or some other memory allocation call prior to registering the memory.  After calling register, it cannot call free(), munmap(), cudaFree(), etc. without first unregistering the region.

If FI_MR_BIT_5 is unset (0), the above constraints are lifted.  You seem to be asking to expand the meaning to include:

- The **app** can change the VA to PA mapping at any time, but only within the same memory domain.  That is, host pages may be replaced with other host pages.
- Or the bit only applies to host memory regions.

Basically, the VA to PA can change, but you don't want that change to mean that the pages can switch between the host and a device.

And we need to be clear here.  The VA to PA mapping is completely outside the control of the app.  The OS can change the mapping at any time.  The restriction on the app is that it won't invoke calls that make this likely to happen.  It's that restriction that enables the provider to work with the OS to avoid it changing the mapping.  Meaning, the provider pins the memory, so the PA is protected from being reused.  The app side of the bargain is that it will keep its VA valid.

Now, with **automatic** page migration to/from device memory, you want to expose this as:
```
FI_MR_BIT_12
When set, the app does... **what**?
```
The point of exposing this flag to the app is for the app to dynamically change its execution.  I have no idea what the app is supposed to do differently.

From the viewpoint of the app, page pinning and ODP aren't relevant.  What matters is the **behavior** that the app is supposed to comply with.

**Reply by iziemba** [reply]
>The action here is that the app must call malloc(), calloc(), mmap(), cudaMalloc(), or some other memory allocation call prior to registering the memory. After calling register, it cannot call free(), munmap(), cudaFree(), etc. without first unregistering the region.

Yes. I would like to see FI_MR_ALLOCATED = 1 be updated with this information. This will make it clear to apps that if this is not set, I can register large VA range and mmap/munmap within it and the MR is still valid. This addresses one of HPEs use cases.

If FI_MR_ALLOCATED = 0, this would mean provider can support VA to PA changes. This opens the door for page migration not only within the same memory domain, but across memory domains. Seems like we could reuse FI_MR_ALLOCATED = 0 for migration, correct? If so, I do not see a need for FI_MR_BIT_12.

In practice, most DMA providers will probably hook into Linux kernel HMM to implement migration. While HMM should always work for FI_HMEM_SYSTEM, whether or not it works for other devices is vendor hardware and software dependent. Thinking out loud, if we had a bit mask denoting for which devices FI_HMEM_SYSTEM can migrate to, apps can change their behavior. For example, if I know memory can migrate to FI_HMEM_CUDA, I'll use the malloc() VA in my GPU kernel. If not, I know I need to allocate GPU device memory.

**Reply by shefty** [reply]
I'm fine clarifying the definition of FI_MR_ALLOCATED.  That's not a problem as long as it follows its original intent.

I'm even okay having some other bit indicate something related to device migration.  But I don't know how to define such a bit.  HMM migrates the pages without the app being aware, so I don't know how the app can take steps to prevent it.  Maybe it's as simple as saying that all device memory must use some GPU specific function, and migration will never automatically occur.

**Reply by iziemba** [reply]
The problem here is some HW platforms may or may not support host to GPU migration with system memory. How is an app suppose to know this? Framing the question like this leads me to think that this is not a libfabric problem if we state that FI_MR_ALLOCATED = 0 allows system migration. Where this memory migrates to may not be a provider concern. For example, TCP provider should still be able to reference the VA and OS takes care of any migration. Similar with with verbs and CXI with ODP and HMM. If this is the case, we may not need a libfabric device migration bit either.

**Comment by iziemba** [comment]
I have opened up https://github.com/ofiwg/libfabric/pull/10062 for this.
